src/net/transport.rs

```diff
@@ -0,0 +1,331 @@
+use crate::{
+    helpers::{
+        query::QueryCommand,
+        transport::{SubscriptionType, Transport, TransportCommand, TransportError},
+        CommandEnvelope, HelperIdentity,
+    },
+    net::{
+        client::MpcHelperClient,
+        discovery::peer,
+        server::{BindTarget, MpcHelperServer},
+    },
+    protocol::QueryId,
+    sync::{Arc, Mutex},
+    task::JoinHandle,
+};
+use async_trait::async_trait;
+use std::collections::{hash_map::Entry, HashMap};
+use std::net::SocketAddr;
+use tokio::sync::mpsc;
+use tokio_stream::wrappers::ReceiverStream;
+
+pub struct HttpTransport {
+    id: HelperIdentity,
+    peers_conf: Arc<[peer::Config; 3]>,
+    subscribe_receiver: Arc<Mutex<Option<mpsc::Receiver<CommandEnvelope>>>>,
+    ongoing_queries: Arc<Mutex<HashMap<QueryId, mpsc::Sender<CommandEnvelope>>>>,
+    server: MpcHelperServer,
+    clients: [MpcHelperClient; 3],
+}
+
+impl HttpTransport {
+    #[must_use]
+    pub fn new(id: HelperIdentity, peers_conf: Arc<[peer::Config; 3]>) -> Arc<Self> {
+        let (subscribe_sender, subscribe_receiver) = mpsc::channel(1);
+        let ongoing_queries = Arc::new(Mutex::new(HashMap::new()));
+        let server = MpcHelperServer::new(subscribe_sender, Arc::clone(&ongoing_queries));
+        let clients = MpcHelperClient::from_conf(&peers_conf);
+        Arc::new(Self {
+            id,
+            peers_conf,
+            subscribe_receiver: Arc::new(Mutex::new(Some(subscribe_receiver))),
+            ongoing_queries,
+            server,
+            clients,
+        })
+    }
+
+    /// Binds self to port described in `peers_conf`.
+    /// # Panics
+    /// if self id not found in `peers_conf`
+    pub async fn bind(&self) -> (SocketAddr, JoinHandle<()>) {
+        let this_conf = &self.peers_conf[self.id];
+        let port = this_conf.origin.port().unwrap();
+        let target = BindTarget::Http(format!("127.0.0.1:{}", port.as_str()).parse().unwrap());
```

Comment:
```suggestion
        let target = BindTarget::Http(format!("0.0.0.0:{}", port.as_str()).parse().unwrap());
```

We should avoid binding to loopback.